controllers/progressivesync_controller.go

```diff
@@ -341,168 +281,265 @@ func (r *ProgressiveSyncReconciler) getOwnedAppsFromClusters(ctx context.Context
+	}
+
+	// To be on the safe side, the controller looks at ALL the progressing Applications selected by the stage selector.
+	// This to avoid having too many Applications in progress triggered by an external process.
+	progressingApps := utils.GetAppsByHealthStatusCode(syncedApps, health.HealthStatusProgressing)
 
-		r.updateStageStatus(ctx, stage.Name, message, syncv1alpha1.PhaseFailed, &ps)
+	// If we reached the maximum number of progressing apps for the stage
+	// then the stage is progressing
+	if len(progressingApps) >= maxTargets {
+		return syncv1alpha1.StageStatusProgressing, nil
+	}
 
-		failed := ps.NewStatusCondition(syncv1alpha1.CompletedCondition, metav1.ConditionFalse, syncv1alpha1.StagesFailedReason, message)
-		apimeta.SetStatusCondition(ps.GetStatusConditions(), failed)
+	outOfSyncApps := utils.GetAppsBySyncStatusCode(selectedApps, argov1alpha1.SyncStatusCodeOutOfSync)
 
-		log.Info("sync failed")
-		return ps, ctrl.Result{Requeue: true, RequeueAfter: RequeueDelayOnError}, nil
+	// If there are no out-of-sync apps then the stage is completed
+	if len(outOfSyncApps) == 0 {
+		return syncv1alpha1.StageStatusCompleted, nil
 	}
 
-	if scheduler.IsStageInProgress(apps, stage, pss) {
-		message := fmt.Sprintf("%s stage in progress", stage.Name)
-		log.Info(message)
+	// If the number of apps synced in the current stage
+	// is equal or greater than the maximum number of targets to sync,
+	// there is nothing else to do.
+	if len(syncedInCurrentStage) >= maxTargets {
+		return syncv1alpha1.StageStatusCompleted, nil
+	}
 
-		for _, app := range apps {
-			log.Info("application details", "app", fmt.Sprintf("%s/%s", app.Namespace, app.Name), "sync.status", app.Status.Sync.Status, "health.status", app.Status.Health.Status)
-		}
+	log.Info("apps summary", "progressingApps", utils.GetAppsName(progressingApps), "syncedInCurrentStage", utils.GetAppsName(syncedInCurrentStage), "OutOfSync", utils.GetAppsName(outOfSyncApps), "stage", stage.Name)
 
-		r.updateStageStatus(ctx, stage.Name, message, syncv1alpha1.PhaseProgressing, &ps)
+	// If there is an external process triggering a sync,
+	// maxParallel - len(progressingApps) might actually be greater than len(outOfSyncApps)
+	// causing the runtime to panic
+	maxParallel = maxParallel - len(progressingApps)
+	if maxParallel > len(outOfSyncApps) {
+		maxParallel = len(outOfSyncApps)
+	}
 
-		progress := ps.NewStatusCondition(syncv1alpha1.CompletedCondition, metav1.ConditionFalse, syncv1alpha1.StagesProgressingReason, message)
-		apimeta.SetStatusCondition(ps.GetStatusConditions(), progress)
+	// Consider the following scenario
+	//
+	// maxTargets = 3
+	// maxParallel = 3
+	// outOfSyncApps = 4
+	// syncedInCurrentStage = 2
+	// progressingApps = 1
+	//
+	// Without the following logic we would end up
+	// with a total of 4 applications synced in the stage
+	if maxParallel+len(syncedInCurrentStage) > maxTargets {
```

Comment:
the logic seems exactly the same as below, but it's a bit confusing the way it's presented.
I think we should refactor it with the lines above.
```
currentParallel := min(maxParallel - len(progressingApps), len(outOfSync), maxTargets - len(syncedInCurrentStage)
```

I think it would make the logic clearer.

Reply:
Before the code review with @Smirl I was using a variable called `maxSync` and the code was
```
	maxSync := maxParallel - len(progressingApps)
	if maxSync > len(outOfSyncApps) {
		maxSync = len(outOfSyncApps)
	}
```
Since we were using maxParallel only to calculate maxSync, @Smirl suggested to overwrite maxParallel to avoid adding a new variable.

I'm happy with both approach to be honest, but I would probably like to keep the calculation explicit rather than a 1 liner because I found it easier to understand.

@Smirl what do you think?

Reply:
💯  on the explicit, but the current calculation doesn't make it obvious that you're basically trying to calculate a low water mark through 3 values and it took me a while to read through it in views with lines 470-473

Reply:
So, should we go back to the previous code with `maxSync`?

Reply:
Pushed